### DISCUSSION TITLE
Confusion around ranges

### DISCUSSION THREAD

**Discussion by Or1g3n** [nushell/nushell#14099] [General]
It is confusing why only the first command works while the second command does not:
1. `0..5 | $in | range (-2)..`
2. `0..5 | range (-2)..`

This is confusing because `[0,1,2,3,4,5] | range (-2)` also works.

I then analyzed the types of each input to see if .. | range (n) only supports a certain input type. Performing `[0,1,2,3,4,5] | describe` returns type list\<int\> so I thought perhaps the range subset command only works against this type. However, both `0..5 | $in | describe` and `0..5 | describe` return range types. So why would `0..5 | $in | range (-2)..` work while `0..5 | range (-2)..` does not.

**Comment by ghost** [comment]
`0..5` is a stream, when using `| $in` it gets collected, `| describe` either for some reason collects it or does not support displaying it in this case.

you can analyse this:

```
$ view ir {|| 0..5 }
# 4 registers, 5 instructions, 0 bytes of data
   0: load-literal           %1, int(0)
   1: load-literal           %2, nothing
   2: load-literal           %3, int(5)
   3: load-literal           %0, range(%1, %2, %3, Inclusive)
   4: return                 %0

$ view ir {|| 0..5 | $in }
# 4 registers, 10 instructions, 0 bytes of data
   0: load-literal           %1, int(0)
   1: load-literal           %2, nothing
   2: load-literal           %3, int(5)
   3: load-literal           %0, range(%1, %2, %3, Inclusive)
   4: collect                %0
   5: clone                  %1, %0
   6: store-variable         var 576, %1
   7: load-variable          %0, var 576
   8: drop-variable          var 576
   9: return                 %0
```

i assume the `Command does not support range input.` either collects it before generating the error or dosn't support giving the info that it is a stream.

But i agree, that either a fix or improvement to the error-message should probably be implemented.

**Reply by fdncred** [reply]
Agreed. Good explanation. Also agree that a better error message in this situation would be helpful, You can also get a result with this `0..5 | collect | range (-2)..`

**Reply by Or1g3n** [reply]
Thanks @Jan9103, that definitely helped. And thanks @fdncred for reviewing this as well.

Given the above, unless I am misunderstanding, it seems unintuitive that a range command in NuShell would not be compatible with handling a range type input. Would it make sense for the range slicing command to auto collect if passed type range (stream of numerics)? Will trust your judgement on this as I am no expert but trying to learn and this is a bit confusing.

**Reply by fdncred** [reply]
I'd prefer to make collection explicit vs implicit. If it's implicitly collecting, there is no command to really change that behavior and make it stream again. If it's explicitly collecting, that means you're providing $in or collect so at least you can change the behavior to what you want.

**Reply by Or1g3n** [reply]
Thanks, that does make sense!
